internal/executor/instance/persistentworker/isolation/none/none.go

```diff
@@ -74,7 +89,45 @@ func (pwi *PersistentWorkerInstance) Run(ctx context.Context, config *runconfig.
 	}
 
 	// Run the agent
-	return cmd.Run()
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// Create a completely separate context <>
+	runCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			pwi.logger.Debugf("gracefully terminating agent with PID %d", cmd.Process.Pid)
+			_ = cmd.Process.Signal(os.Interrupt)
+		case <-runCtx.Done():
+			// agent exited
+			return
+		}
+
+		const gracefulLimit = 3 * time.Second
```

Comment:
Do you think it's a bit too low? Maybe give it, let's say, 30 seconds?

Reply:
I believe it's already high enough since we're somewhat violating the `context.Context` expectations of an immediate termination in `Run()`.

Plus, technically it's possible to terminate uncleanly even with a 30 second timeout. Perhaps we should log these cases instead and analyze them to see if it's actually a problem?

The other option is pass a file name of the log file before it's created from the agent and ensure it's deleted from the CLI later.

Reply:
I'm just thinking about potential cases of executing Gradle daemon, mysql, etc. Some heavy things might take a while to get terminalted. Agreed it's not a big issue. I just was curious about the 3 seconds constant and thought it's a bit low. Will leave it up to your judgement then. 👍

Reply:
> I'm just thinking about potential cases of executing Gradle daemon, mysql, etc. Some heavy things might take a while to get terminalted.

Note that these things will be killed (https://github.com/cirruslabs/cirrus-ci-agent/blob/061ab28d68561215957f6c92752ada0dcd55369b/internal/executor/shell.go#L65), not gracefully terminated.